odoo/tools/template_inheritance.py

```diff
@@ -205,23 +205,94 @@ def extract(spec):
+                                            break
+                                        expression = part.group(0)
+                                        start_parentesis = part.group(1)
+                                        to_remove = part.group(3)
+                                        end_parentesis = part.group(4)
+
+                                        start_n = start_parentesis.split(' (')
+                                        end_n = end_parentesis.split(') ')
+                                        if len(start_n) > len(end_n):
+                                            expression = ' ('.join(start_n[-len(end_n):]) + to_remove + end_parentesis
+                                        elif len(start_n) < len(end_n):
+                                            expression = start_parentesis + to_remove + ') '.join(end_n[0:len(start_n)])
+
+                                        if not is_constant and ' (' not in expression:
+                                            break
+
+                                        if expression == value:
+                                            value = ''
+                                            continue
+                                        elif f' {separator}{expression}' in value:
+                                            value = value.replace(f' {separator}{expression}', ' ')
+                                        elif f'{expression}{separator} ' in value:
+                                            value = value.replace(f'{expression}{separator} ', ' ')
+                                        else:
+                                            break
+
+                                    if value == ' ':
+                                        value = ''
+
+                            if child.get('add'):
+                                add = child.attrib.pop('add')
+                                value = f"{value}{separator} ({add}) " if value else add
+
+                        else:
+                            separator = child.attrib.pop('separator', None)
+                            if separator is None:
+                                separator = ','
+                            if separator == ' ':
+                                separator = None    # squash spaces
+                            to_add = (
+                                s for s in (s.strip() for s in child.attrib.pop('add', '').split(separator))
+                                if s
+                            )
+                            to_remove = {s.strip() for s in child.attrib.pop('remove', '').split(separator)}
+                            values = (s.strip() for s in node.get(attribute, '').split(separator))
+                            value = (separator or ' ').join(itertools.chain(
+                                (v for v in values if v not in to_remove),
+                                to_add
+                            ))
```

Comment:
The existing code looks awfully complicated to me.  Why don't you make an ordered set with the `value`?
```suggestion
                            items = OrderedSet(s.strip() for s in node.get(attribute, '').split(separator))
                            for s in filter(None, child.attrib.pop('remove', '').split(separator)):
                                items.discard(s.strip())
                            for s in filter(None, child.attrib.pop('add', '').split(separator)):
                                items.add(s.strip())
                            value = (separator or ' ').join(items)
```